service/debugger/debugger.go

```diff
@@ -1475,6 +1487,158 @@ func (d *Debugger) Functions(filter string) ([]string, error) {
+	}
+	return true
+}
+
+func dequeue(queue *list.List) TraceFuncptr {
+	element := queue.Front()
+	queue.Remove(element)
+	data := element.Value
+	F, ok := data.(TraceFuncptr)
+	if !ok {
+		fmt.Printf("Type error\n")
+		os.Exit(1)
+	}
+	return F
+}
+
+func traverse(t proc.ValidTargets, f *proc.Function, depth int, FollowCalls int) ([]string, error) {
+
+	var queue *list.List
+	queue = list.New()
+	queue.Init()
+
+	funcs := []string{}
+
+	rootnode := &TraceFunc{Func: new(proc.Function), Depth: depth, children: []*proc.Function{}, visited: false, dis: false}
+	rootnode.Func = f
+
+	TraceMap[f.Name] = rootnode
+	queue.PushBack(TraceMap[f.Name])
+	for queue.Len() > 0 {
+		parent := dequeue(queue)
+		if parent.Depth > FollowCalls {
+			continue
+		}
+		if !parent.visited {
+			// Guard to avoid adding duplicates to funcs
+			funcs = append(funcs, parent.Func.Name)
+			parent.visited = true
+		}
+		if parent.Depth+1 > FollowCalls {
+			// Skip disassembling and computing children if we are going to cross FollowCalls
+			continue
+		}
+		deferdepth := 0
+		var deferparent TraceFuncptr = nil
+		if !parent.dis {
+			f := parent.Func
+			text, err := proc.Disassemble(t.Memory(), nil, t.Breakpoints(), t.BinInfo(), f.Entry, f.End)
+			if err != nil {
+				return nil, fmt.Errorf("disassemble failed with error %w", err)
+			}
+			for _, instr := range text {
+				if instr.IsCall() && instr.DestLoc != nil && instr.DestLoc.Fn != nil {
+					cf := instr.DestLoc.Fn
+					if !filter(cf.Name) {
+						continue
+					}
+					childnode := TraceMap[cf.Name]
+					if childnode == nil {
+						childnode = &TraceFunc{Func: new(proc.Function), Depth: parent.Depth + 1, children: []*proc.Function{}, visited: false, dis: false}
+						childnode.Func = cf
+						TraceMap[cf.Name] = childnode
+						parent.children = append(parent.children, cf)
+					} else {
+						if childnode.Depth > parent.Depth+1 {
+							// can go even further deep
+							childnode.Depth = parent.Depth + 1
+							parent.children = append(parent.children, cf)
+						}
+					}
+
+					if strings.Contains(cf.Name, "runtime.deferreturn") {
+						if deferparent == nil && childnode.Depth+1 <= FollowCalls {
+							// keep track of defer function inorder to trace defer calls
+							deferdepth = childnode.Depth
+							deferparent = parent
+						}
+					}
+
+				}
+			}
+			parent.dis = true
+		}
+		for _, cf := range parent.children {
+			if cf.Name == parent.Func.Name {
+				//recursive functions
+				continue
+			}
+			childnode := TraceMap[cf.Name]
+			if childnode == nil {
+				// childnodes should have already been computed in the above loop
+				fmt.Printf("unexpected error child nil \n")
+				os.Exit(1)
+			}
+			if childnode.Depth < parent.Depth+1 {
+				// This path has already been explored no need to revisit
+				continue
+			}
+			if childnode.Depth <= FollowCalls {
+				queue.PushBack(TraceMap[cf.Name])
+			}
+		}
+
+		if deferdepth > 0 {
+			for _, fbinary := range t.BinInfo().Functions {
```

Comment:
We can actually get rid of this function name lookup heuristic. This will only work when deferring anonymous functions such as:

```
defer func() {}()
```

What we should actually be doing is looking for an assembly sequence such as:

```
lea rcx, ptr [rip+0x24ab3]
mov qword ptr [rsp+0x20], rcx
```

Which is loading into a register (in this case `$rcx`) the address of the deferred function, which it is referencing as an offset from the current instruction pointer. It then will eventually load this into a stack slot. You can calculate what the instruction pointer value will be with some arithmetic. Essentially, it would be `image.StaticBase + (function entry address) + (offset of this instruction, the number of bytes decoded so far)`.

It's necessary to take the `image.StaticBase` into account for handling PIE binaries. On non-PIE binaries that value will be 0 anyways.

Reply:
Note that also this address may point to a defer wrapper function on certain Go versions, but that's ok for now, we can skip past the wrapper function later, or not increment depth if we hit the wrapper function so that we effectively skip it.

Reply:
as discussed the support for defer calls in the follow funcs feature will be done in a follow up PR